### DISCUSSION TITLE
Harbison book - Modula-3

### DISCUSSION THREAD

**Discussion by dukester** [modula3/cm3#1158] [General]
Found it on Internet Archive.
That book is the ticket!! I'll be looking around locally for it. I sure that I can stumble onto a copy. The Internet Archive copy is read-only.

Is this book what some of you have referred to as "the green book"?

**Comment by mikanystrom-intel** [comment]
Harbison is a great introduction to Modula-3, but no it's not "the green book".  Harbison is white :-)

The green book is "Systems Programming with Modula-3", edited by Greg Nelson, published by Prentice-Hall.  Someone, Olaf I think, just posted a link to a soft copy:

https://archive.org/details/systems-programming-with-modula-3/page/x/mode/2up

**Reply by dukester** [reply]
Right! I've got Nelson's book! It sits heavy on me at this stage. Harbison's is what I'm going after. Thanks for clarifying the colors! 😄

**Reply by mikanystrom-intel** [reply]
Yes the green book is not exactly lightweight.  I had an MS in CS when I read it the first time, and it took me a few readings to appreciate the brilliance of it.  Well, no that's not exactly true, that there was something brilliant going on was obvious to me on the first reading, but I didn't really get how it all hung together so well.  How can you get something that's at the same time so practical and so elegant with so little fuss?  Yes there is SOME fuss in Modula-3..(BRANDED for example) but it's not public-static-void-main fussy.

**Reply by dukester** [reply]
It makes a person wonder why such a brilliant language failed to become more pervasive. Licensing cost back then? It's free now, yet still not adapted. Maybe its brilliance is blinding. At any rate, Modula-3 is hands down more appealing to me than PL/I course I took at McGill U. using a time-share VAX mainframe - punched cards and all. 😄

**Reply by mtbc** [reply]
My guess is that its adoption was impeded partly because it looks like PASCAL instead of looking like C++ which at least seemed to bother some people. Especially before Java came out, Modula-3 answered a real pain point for how to create portable, multi-threaded, graphical applications. (And then when Java 1.0 came out, one read the spec and wept as it had failed to learn so much from M3! Some of which deficiencies they've since fixed years later.)

**Reply by dukester** [reply]
I suppose that the old adage that there's no accounting for people's tastes/choices seem to hold true here. I never could stand c++ !! Once I got my head around pointer indirection, I had very little issues with C. I liked Oberon-2. Maybe Wirth just irritated a whack of folks with his closed Oberon System et al, and anything Wirth-ian was doomed in the long haul. Who knows! DEC's demise might not have helped matters either I suppose.
